![image info](https://raw.githubusercontent.com/albahnsen/MIAD_ML_and_NLP/main/images/banner_1.png)

# Proyecto 2 - Clasificación de género de películas

El propósito de este proyecto es que puedan poner en práctica, en sus respectivos grupos de trabajo, sus conocimientos sobre técnicas de preprocesamiento, modelos predictivos de NLP, y la disponibilización de modelos. Para su desarrollo tengan en cuenta las instrucciones dadas en la "Guía del proyecto 2: Clasificación de género de películas"

**Entrega**: La entrega del proyecto deberán realizarla durante la semana 8. Sin embargo, es importante que avancen en la semana 7 en el modelado del problema y en parte del informe, tal y como se les indicó en la guía.

Para hacer la entrega, deberán adjuntar el informe autocontenido en PDF a la actividad de entrega del proyecto que encontrarán en la semana 8, y subir el archivo de predicciones a la [competencia de Kaggle](https://www.kaggle.com/t/2c54d005f76747fe83f77fbf8b3ec232).

## Datos para la predicción de género en películas

![image info](https://raw.githubusercontent.com/albahnsen/MIAD_ML_and_NLP/main/images/moviegenre.png)

En este proyecto se usará un conjunto de datos de géneros de películas. Cada observación contiene el título de una película, su año de lanzamiento, la sinopsis o plot de la película (resumen de la trama) y los géneros a los que pertenece (una película puede pertenercer a más de un género). Por ejemplo:
- Título: 'How to Be a Serial Killer'
- Plot: 'A serial killer decides to teach the secrets of his satisfying career to a video store clerk.'
- Generos: 'Comedy', 'Crime', 'Horror'

La idea es que usen estos datos para predecir la probabilidad de que una película pertenezca, dada la sinopsis, a cada uno de los géneros.

Agradecemos al profesor Fabio González, Ph.D. y a su alumno John Arevalo por proporcionar este conjunto de datos. Ver https://arxiv.org/abs/1702.01992

## Ejemplo predicción conjunto de test para envío a Kaggle
En esta sección encontrarán el formato en el que deben guardar los resultados de la predicción para que puedan subirlos a la competencia en Kaggle.

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
%pip install livelossplot

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [99]:
# Importación librerías
import os
import re
import nltk
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from sklearn.preprocessing import StandardScaler, MultiLabelBinarizer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.multiclass import OneVsRestClassifier
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.metrics import r2_score, roc_auc_score
from xgboost import XGBClassifier
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras.wrappers.scikit_learn import KerasRegressor
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras import backend as K
from livelossplot import PlotLossesKeras

In [116]:
# Carga de datos de archivo .csv
dataTraining = pd.read_csv('https://github.com/albahnsen/MIAD_ML_and_NLP/raw/main/datasets/dataTraining.zip', encoding='UTF-8', index_col=0)
dataTesting = pd.read_csv('https://github.com/albahnsen/MIAD_ML_and_NLP/raw/main/datasets/dataTesting.zip', encoding='UTF-8', index_col=0)

In [117]:
# Visualización datos de entrenamiento
dataTraining.head()

year                      title  \
3107  2003                       Most   
900   2008  How to Be a Serial Killer   
6724  1941             A Woman's Face   
4704  1954            Executive Suite   
2582  1990              Narrow Margin   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     plot  \
3107                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 most is the story of a single father who takes his eight year - old son to work with him at the railroad drawbridge where he is the bridge tender .  a day before ,  the boy meets a woman boarding a train ,  a drug abuser .  at the bridge ,  the father goes into the engine room ,  and tells his son to stay at the edge of the nearby lake .  a ship comes ,  and the bridge is lifted .  though it is supposed to arrive an hour later ,  the train happens to arrive .  the son sees this ,  and tries to warn his father ,  who is not able to see this .  just as the oncoming train approaches ,  his son falls into the drawbridge gear works while attempting to lower the bridge ,  leaving the father with a horrific choice .  the father then lowers the bridge ,  the gears crushing the boy .  the people in the train are completely oblivious to the fa

In [118]:
# Visualización datos de test
dataTesting.head()

year                title  \
1  1999  Message in a Bottle   
4  1978     Midnight Express   
5  1996          Primal Fear   
6  1950               Crisis   
7  1959          The Tingler   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     plot  
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     who meets by fate ,  shall be sealed by fate .  theresa osborne is running along the beach when she stumbles upon a bottle washed up on the shore .  inside is a message ,  reading the letter s

In [119]:
# Definición de variables predictoras (X)
vect = CountVectorizer(max_features=1000)
X_dtm = vect.fit_transform(dataTraining['plot'])
X_dtm.shape

(7895, 1000)

In [120]:
# Definición de variable de interés (y)
dataTraining['genres'] = dataTraining['genres'].map(lambda x: eval(x))
le = MultiLabelBinarizer()
y_genres = le.fit_transform(dataTraining['genres'])

In [121]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler(with_mean=False)
scaler.fit(X_dtm)
X_train, X_test, y_train_genres, y_test_genres = train_test_split(X_dtm, y_genres, test_size=0.33, random_state=42)
#X_train = pd.DataFrame(data=scaler.transform(X_train))
#X_test = pd.DataFrame(data=scaler.transform(X_test))

In [56]:
# Definición y entrenamiento
clf = OneVsRestClassifier(RandomForestClassifier(n_jobs=-1, n_estimators=100, max_depth=10, random_state=42))
clf.fit(X_train, y_train_genres)

OneVsRestClassifier(estimator=RandomForestClassifier(max_depth=10, n_jobs=-1,
                                                     random_state=42))

In [57]:
# Predicción del modelo de clasificación
y_pred_genres = clf.predict_proba(X_test)

# Impresión del desempeño del modelo
roc_auc_score(y_test_genres, y_pred_genres, average='macro')

0.7812262183677007

In [58]:
y_test_genres

array([[0, 0, 0, ..., 1, 0, 0],
       [0, 1, 0, ..., 0, 1, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 1, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [59]:
# transformación variables predictoras X del conjunto de test
X_test_dtm = vect.transform(dataTesting['plot'])

cols = ['p_Action', 'p_Adventure', 'p_Animation', 'p_Biography', 'p_Comedy', 'p_Crime', 'p_Documentary', 'p_Drama', 'p_Family',
        'p_Fantasy', 'p_Film-Noir', 'p_History', 'p_Horror', 'p_Music', 'p_Musical', 'p_Mystery', 'p_News', 'p_Romance',
        'p_Sci-Fi', 'p_Short', 'p_Sport', 'p_Thriller', 'p_War', 'p_Western']

# Predicción del conjunto de test
y_pred_test_genres = clf.predict_proba(X_test_dtm)


In [60]:
# Guardar predicciones en formato exigido en la competencia de kaggle

res = pd.DataFrame(y_pred_test_genres, index=dataTesting.index, columns=cols)
res.to_csv('pred_genres_text_RF.csv', index_label='ID')
res.head()

p_Action  p_Adventure  p_Animation  p_Biography  p_Comedy   p_Crime  \
1  0.143030     0.101960     0.024454     0.029938  0.354552  0.138830   
4  0.122624     0.085786     0.024213     0.084795  0.370949  0.216657   
5  0.151364     0.110284     0.013762     0.075334  0.304837  0.448736   
6  0.154448     0.125772     0.020991     0.064124  0.340779  0.140892   
7  0.175143     0.210069     0.035476     0.032505  0.313850  0.243150   

   p_Documentary   p_Drama  p_Family  p_Fantasy  ...  p_Musical  p_Mystery  \
1       0.030787  0.490140  0.073159   0.101339  ...   0.025069   0.063208   
4       0.080359  0.515684  0.062976   0.067019  ...   0.024734   0.060935   
5       0.021010  0.611544  0.081741   0.169121  ...   0.044538   0.261372   
6       0.009133  0.632038  0.068287   0.063631  ...   0.131074   0.088418   
7       0.021793  0.427885  0.079781   0.143879  ...   0.023859   0.090359   

     p_News  p_Romance  p_Sci-Fi   p_Short   p_Sport  p_Thriller     p_War  \
1  0.000000   0.362818  0.056648  0.008970  0.017522    0.202605  0.033989   
4  0.000477   0.149703  0.058190  0.014248  0.020099    0.204794  0.030438   
5  0.000000   0.335987  0.128505  0.001016  0.048658    0.423242  0.052693   
6  0.000000   0.197224  0.132208  0.001432  0.039743    0.269385  0.077607   
7  0.000048   0.205117  0.241663  0.002634  0.018403    0.259465  0.021569   

   p_Western  
1   0.018117  
4   0.018506  
5   0.025351  
6   0.017862  
7   0.017585  

[5 rows x 24 columns]

# Desarrollo Proyecto

In [67]:

output_var = y_train_genres.shape[1]
print(output_var, ' output variables')
dims = X_train.shape[1]
print(dims, 'input variables')
dense_shape = X_train.shape
indices = np.column_stack(X_train.nonzero())
values = X_train.data
X_train_sparse = tf.sparse.SparseTensor(
    indices=indices,
    values=values,
    dense_shape=dense_shape
)
X_train_sparse_ordered = tf.sparse.reorder(X_train_sparse)
X_train_dense = tf.sparse.to_dense(X_train_sparse_ordered).numpy()
X_train2, X_val, Y_train2, Y_val = train_test_split(X_train_dense, y_train_genres, test_size=0.15, random_state=42)
def nn_model_params(optimizer ,
                    neurons,
                    batch_size,
                    epochs,
                    activation,
                    patience,
                    loss):
    
    K.clear_session()

    # Definición red neuronal con la función Sequential()
    model = Sequential()
    
    # Definición de las capas de la red con el número de neuronas y la función de activación definidos en la función nn_model_params
    model.add(Dense(neurons, input_shape=(dims,), activation=activation))
    model.add(Dense(neurons, activation=activation))
    model.add(Dense(output_var, activation=activation))

    # Definición de función de perdida con parámetros definidos en la función nn_model_params
    model.compile(optimizer = optimizer, loss=loss)
    
    # Definición de la función EarlyStopping con parámetro definido en la función nn_model_params
    early_stopping = EarlyStopping(monitor="val_loss", patience = patience)

    # Entrenamiento de la red neuronal con parámetros definidos en la función nn_model_params
    model.fit(X_train2, Y_train2,
              validation_data = (X_val, Y_val),
              epochs=epochs,
              batch_size=batch_size,
              callbacks=[early_stopping, PlotLossesKeras()],
              verbose=True
              )
     
    return model

24  output variables
1000 input variables


In [68]:
nn_params = {
    'optimizer': ['adam','sgd'],
    'activation': ['relu','sigmoid','softmax'],
    'batch_size': [64,512],
    'neurons':[64,512],
    'epochs':[20,50],
    'patience':[2,5],
    'loss':['mean_squared_error','binary_crossentropy','categorical_crossentropy']
}

## Método busqueda por cuadrícula (Grid Search)

In [69]:



nn_model = KerasRegressor(build_fn=nn_model_params)
gs = GridSearchCV(nn_model, nn_params, cv=3)

gs.fit(X_train2, Y_train2)

print('Los mejores parametros segun Grid Search:', gs.best_params_)

Loss
	training         	 (min:    0.040, max:    0.105, cur:    0.040)
	validation       	 (min:    0.086, max:    0.101, cur:    0.086)
71/71 [==============================] - 1s 7ms/step - loss: 0.0398 - val_loss: 0.0860
Epoch 1/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0394
Epoch 2/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0392
Epoch 3/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0392
Epoch 4/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0389
Epoch 5/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0388
Epoch 6/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0388
Epoch 7/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0389
Epoch 8/50
71/71 [==============================] - 0s 5ms/step - loss: 0.0388
Epoch 9/50
71/71 [==============================] - 0s 4ms/step - loss: 0.0388
Epoch 10/50
71/71 [==============================] - 0s 5ms/step 

In [122]:
model = nn_model_params(optimizer='adam', neurons=512, batch_size=64, epochs=50, activation='relu', patience=5, loss='mean_squared_error')


Loss
	training         	 (min:    0.031, max:    0.096, cur:    0.031)
	validation       	 (min:    0.083, max:    0.088, cur:    0.085)
71/71 [==============================] - 1s 8ms/step - loss: 0.0306 - val_loss: 0.0850


In [123]:
dense_shape = X_test.shape
indices = np.column_stack(X_test.nonzero())
values = X_test.data
X_test_sparse = tf.sparse.SparseTensor(
    indices=indices,
    values=values,
    dense_shape=dense_shape
)
X_test_sparse_ordered = tf.sparse.reorder(X_test_sparse)
X_test_dense = tf.sparse.to_dense(X_test_sparse_ordered).numpy()
Y_predict_neuronal = model.predict(X_test_dense)

In [124]:
roc_auc_score(y_test_genres, Y_predict_neuronal, average='macro')

0.6019548999447989

In [125]:
X_test_dtm = vect.transform(dataTesting['plot'])
dense_shape = X_test_dtm.shape
indices = np.column_stack(X_test_dtm.nonzero())
values = X_test_dtm.data
X_test_dtm_sparse = tf.sparse.SparseTensor(
    indices=indices,
    values=values,
    dense_shape=dense_shape
)
X_test_dtm_sparse_ordered = tf.sparse.reorder(X_test_dtm_sparse)
X_test_dtm_dense = tf.sparse.to_dense(X_test_dtm_sparse_ordered).numpy()
Y_predict_neuronal = model.predict(X_test_dtm_dense)
# Predicción del conjunto de test
res = pd.DataFrame(Y_predict_neuronal, index=dataTesting.index, columns=cols)
res.to_csv('pred_genres_neuronal.csv', index_label='ID')
res.head()

p_Action  p_Adventure  p_Animation  p_Biography  p_Comedy   p_Crime  \
1   0.00000          0.0          0.0          0.0  0.281182  0.000000   
4   0.18873          0.0          0.0          0.0  0.116675  0.144052   
5   0.00000          0.0          0.0          0.0  0.000000  0.165793   
6   0.00000          0.0          0.0          0.0  0.127128  0.000000   
7   0.00000          0.0          0.0          0.0  0.000000  0.000000   

   p_Documentary   p_Drama  p_Family  p_Fantasy  ...  p_Musical  p_Mystery  \
1            0.0  0.426335       0.0   0.000000  ...        0.0     0.0000   
4            0.0  1.014342       0.0   0.000000  ...        0.0     0.0000   
5            0.0  0.919080       0.0   0.000000  ...        0.0     1.0556   
6            0.0  0.481070       0.0   0.000000  ...        0.0     0.0000   
7            0.0  0.477212       0.0   0.322055  ...        0.0     0.0000   

   p_News  p_Romance  p_Sci-Fi  p_Short  p_Sport  p_Thriller  p_War  p_Western  
1     0.0   0.888930  0.000000      0.0      0.0    0.000000    0.0        0.0  
4     0.0   0.000000  0.000000      0.0      0.0    0.000000    0.0        0.0  
5     0.0   0.000000  0.000000      0.0      0.0    1.006275    0.0        0.0  
6     0.0   0.106217  0.000000      0.0      0.0    0.230854    0.0        0.0  
7     0.0   0.166783  0.530611      0.0      0.0    0.016101    0.0        0.0  

[5 rows x 24 columns]

Para este caso se realizo el entrenamiento de una red neuronal en la cual en primer lugar se realizo el estandarizado de los datos con el proposito de que todos tuvieran la misma dimensionalidad para despues  calibrar los hiperparametros usando GridSearchCV en donde se obtuvieron los siguientes parametros : {'activation': 'relu', 'batch_size': 64, 'epochs': 50, 'loss': 'mean_squared_error', 'neurons': 512, 'optimizer': 'adam', 'patience': 5} en donde despues se prosigue a evaluar su poder predictivo en donde nos da un roc de 0.6 en donde el modelo presenta un poder predictivo bajo

## Segundo metodo

In [94]:

nltk.download('stopwords')
nltk.download('wordnet')


[nltk_data] Downloading package stopwords to
[nltk_data]     /home/richard/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /home/richard/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [95]:
dataTraining = pd.read_csv('https://github.com/albahnsen/MIAD_ML_and_NLP/raw/main/datasets/dataTraining.zip', encoding='UTF-8', index_col=0)
dataTesting = pd.read_csv('https://github.com/albahnsen/MIAD_ML_and_NLP/raw/main/datasets/dataTesting.zip', encoding='UTF-8', index_col=0)

# Preprocesamiento de datos
dataTraining['genres'] = dataTraining['genres'].map(lambda x: eval(x))
dataTraining.dropna(subset=['plot'], inplace=True)
dataTesting.dropna(subset=['plot'], inplace=True)


In [96]:
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

def clean_text(text):
    text = re.sub(r'\b\w{1,2}\b', '', text)  # Remover palabras cortas
    text = re.sub(r'[^\w\s]', '', text)  # Remover puntuación
    text = text.lower()
    text = ' '.join([lemmatizer.lemmatize(word) for word in text.split() if word not in stop_words])
    return text

In [98]:


# Descarga los recursos necesarios de NLTK
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')

# Define el conjunto de stopwords y el lematizador
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

# Define la función para limpiar el texto
def clean_text(text):
    # Elimina palabras con menos de 3 caracteres
    text = re.sub(r'\b\w{1,2}\b', '', text)
    # Convierte el texto a minúsculas
    text = text.lower()
    # Elimina caracteres no alfabéticos
    text = re.sub(r'[^a-z\s]', '', text)
    # Elimina stopwords y lematiza las palabras restantes
    text = ' '.join(lemmatizer.lemmatize(word) for word in text.split() if word not in stop_words)
    return text

# Aplica la función clean_text a las columnas 'plot'
dataTraining['clean_plot'] = dataTraining['plot'].apply(clean_text)
dataTesting['clean_plot'] = dataTesting['plot'].apply(clean_text)

# Vectorización del texto (usando TF-IDF)
vect = TfidfVectorizer(max_features=1000)
X_dtm = vect.fit_transform(dataTraining['clean_plot'])

# Añadir característica adicional: año de lanzamiento
X_additional = dataTraining[['year']].values
scaler = StandardScaler()
X_additional = scaler.fit_transform(X_additional)

# Concatenar las características
X = np.hstack((X_dtm.toarray(), X_additional))

# Binariza las etiquetas de género
le = MultiLabelBinarizer()
y_genres = le.fit_transform(dataTraining['genres'])

# División en conjuntos de entrenamiento y prueba
X_train, X_test, y_train_genres, y_test_genres = train_test_split(X, y_genres, test_size=0.33, random_state=42)

# Definición del modelo
clf = OneVsRestClassifier(XGBClassifier(n_jobs=-1, random_state=42))

# Parámetros para GridSearchCV
param_grid = {
    'estimator__max_depth': [3, 5],
    'estimator__n_estimators': [100, 200],
    'estimator__learning_rate': [0.1, 0.01]
}

# Búsqueda de hiperparámetros con GridSearchCV
grid_search = GridSearchCV(clf, param_grid, cv=3, scoring='roc_auc', n_jobs=-1)
grid_search.fit(X_train, y_train_genres)

# Obtener el mejor modelo
best_clf = grid_search.best_estimator_

# Predicción del modelo optimizado
y_pred_genres = best_clf.predict_proba(X_test)

# Evaluación del desempeño del modelo
score = roc_auc_score(y_test_genres, y_pred_genres, average='macro')
print(f'ROC AUC Score: {score:.4f}')

# Transformación de las variables predictoras del conjunto de test
X_test_dtm = vect.transform(dataTesting['clean_plot'])
X_test_additional = scaler.transform(dataTesting[['year']].values)
X_test_final = np.hstack((X_test_dtm.toarray(), X_test_additional))

# Predicción del conjunto de test
y_pred_test_genres = best_clf.predict_proba(X_test_final)

res = pd.DataFrame(y_pred_test_genres, index=dataTesting.index, columns=cols)
res.to_csv('pred_genres_text_optimized.csv', index_label='ID')
res.head()


[nltk_data] Downloading package stopwords to
[nltk_data]     /home/richard/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /home/richard/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /home/richard/nltk_data...


ROC AUC Score: 0.8443


p_Action  p_Adventure  p_Animation  p_Biography  p_Comedy   p_Crime  \
1  0.092402     0.084652     0.012940     0.035859  0.250526  0.063461   
4  0.111633     0.011719     0.007698     0.130590  0.299000  0.373507   
5  0.104538     0.028352     0.001959     0.013698  0.153403  0.895682   
6  0.075964     0.050919     0.001829     0.057074  0.099848  0.080855   
7  0.097777     0.060696     0.005795     0.039584  0.189369  0.046483   

   p_Documentary   p_Drama  p_Family  p_Fantasy  ...  p_Musical  p_Mystery  \
1       0.008995  0.563048  0.038448   0.129786  ...   0.011297   0.059565   
4       0.036538  0.720126  0.012663   0.015727  ...   0.030510   0.049601   
5       0.003289  0.782740  0.003807   0.046089  ...   0.004880   0.703880   
6       0.001344  0.693180  0.037175   0.019650  ...   0.022400   0.131353   
7       0.004461  0.138044  0.093255   0.378920  ...   0.015328   0.032513   

     p_News  p_Romance  p_Sci-Fi   p_Short   p_Sport  p_Thriller     p_War  \
1  0.000027   0.641219  0.049179  0.001708  0.008198    0.249507  0.003604   
4  0.000169   0.048189  0.035659  0.010540  0.014539    0.135603  0.014654   
5  0.000057   0.159485  0.021397  0.000181  0.014431    0.523088  0.016002   
6  0.000028   0.254997  0.052519  0.000857  0.014590    0.257001  0.074340   
7  0.000026   0.042444  0.891800  0.002538  0.004779    0.140894  0.028046   

   p_Western  
1   0.003297  
4   0.011281  
5   0.001365  
6   0.024038  
7   0.028572  

[5 rows x 24 columns]

En este caso se realiza un modelo con XGBClassifier en el cual se obtiene un Roc de 0.84 demostrando un mejor poder predictivo que la red neuronal.